[v3] Notifications API

popaprozac opened the pull request:
# Description

Adds basic notification support to macOS via [UNNotification API](https://developer.apple.com/documentation/usernotifications), Windows via [go-toast](https://pkg.go.dev/git.sr.ht/~jackmordaunt/go-toast/v2@v2.0.3), and currently testing on Linux via dbus.

Fixes (https://github.com/wailsapp/wails/issues/1788)

https://github.com/user-attachments/assets/e2b204f4-3b7e-4328-abf3-587991ee4975

#### Current implementation:
- [x] Basic notifications
- [x] Notification actions
- [x] Notification interaction events
- [x] JS API

#### Future features:
- [ ] Sounds
- [ ] Icons/images/attachments
- [ ] Notification priority
- [ ] Windows Toast `ActivationType` options?
- [ ] Respond to notification events when app is closed (is this possible?)

#### Known changes before:
- [x] Refactor/improve error handling
- [x] Refactor/improve `NotificationResponseData`
- [x] go-toast `AppData` config
- [x] Linux support?
- [x] Add optional metadata to basic notifications (macOS)

#### Limitations/Notes/Thoughts/Questions:
**macOS:** 
- Your app **must** have a bundle identifier and be signed for the notification API to work! I try to call this out everywhere here and in the code but this will make the dev experience tougher. Open to suggestions.
- Minimum version supported macOS 11.0

**Windows:**
- To match macOS's ability to add arbitrary data to notifications I am encoding that data and adding it to the button "arguments" and decoding it back out when interacted with. Don't love it but works.
- ~Should we include the icon by default to the toast config? macOS provides the app icon by default in the notification banner.~
- ~Need to work to better understand Windows app GUIDs for go-toast `AppData` config~

**Linux:**
- Tested on Fedora, Ubuntu, and Debian. More importantly is testing desktop environments. Tested on GNOME, MATE, LXDE, Cinnamon, Plasma...when I get the energy I will outline the minor differences but all work!
- ~Taking inspiration from prior attempt [here](https://github.com/wailsapp/wails/pull/2206/files#diff-f582b98e814e4fcf8ebe92bca528d8db4395a40726da81ef3953cfb71c0f5edc). Would love some help!~
- ~Look at [shout](https://git.sr.ht/~whereswaldon/shout) to better understand what this will take, would love help~

#### Platform differences:
| Feature | macOS | Windows | Linux |
|---|---|---|---|
| Permissions | macOS requires explicit user permission to send notifications from an application.<br>The app requires a valid bundle identifier and needs to be signed.<br><br>Currently the dev build or binary will not launch and warn in the console. We probably want this to not block but just warn with no-op to any calls?<br>I am going to slightly refactor `CheckBundleIdentifier` since original source for this implementation logged from Obj-C. No need, we can handle in Go.<br><br>`CheckBundleIdentifier`, `RequestUserNotificationAuthorization`, and `CheckNotificationAuthorization` are available on macOS.<br><br>~I want to do more testing since I had to hack together a timeout to wait for `CheckNotificationAuthorization`.~<br>~Might just reuse `RequestUserNotificationAuthorization` if I can.~ | `CheckBundleIdentifier`, `RequestUserNotificationAuthorization`, and `CheckNotificationAuthorization` always returns `true`. | `CheckBundleIdentifier`, `RequestUserNotificationAuthorization`, and `CheckNotificationAuthorization` always returns `true`. |
| Notifications | Supports:<br><ul><li>[x] *Identifier</li><li>[x] Title</li><li>[x] Subtitle</li><li>[x] Body</li><li>[x] Actions</li><li>[x] Text Input</li><li>[x] Custom data</li></ul><br>*Identifier should be unique and can be used by other methods to update or remove a specific notification.<br>`Destructive` action option is available on macOS. | Supports:<br><ul><li>[x] *Identifier</li><li>[x] Title</li><li>[ ] Subtitle</li><li>[x] Body</li><li>[x] Actions</li><li>[x] Text Input</li><li>[x] Custom data</li></ul><br>*Identifier is the toast's "AppID" and presented like a notification name.<br>`Destructive` action option is not available. | Supports:<br><ul><li>[x] *Identifier</li><li>[x] Title</li><li>[ ] Subtitle</li><li>[x] Body</li><li>[x] Actions</li><li>[x] Text Input</li><li>[x] Custom data</li></ul><br>*Identifier should be unique and can be used by other methods to update or remove a specific notification.<br>`Destructive` action option is not available. |
| Notification Categories | For more complex notifications you can register a notification category with specified actions and input options.<br><br>Registered notification categories persist for the application, even between unauthorization and reauthorization of notifications for the app. This is handled by `UNUserNotificationCenter` API.<br><br>Registering needs to only occur once per unique category ID. Reregistering overrides the previous config. `RemoveNotificationCategory` is there if needed.<br><br>For simplicity, only a single text field is currently configurable. Maybe make name more generic? Currently referred to as a reply field. If set to true you can customize the text input placeholder and button. | Implemented matching API that registers notification categories in the Registry on Windows.<br><br>On service start and shutdown the registry is updated.<br><br>To match the macOS implementation if reply field is set to true a single text input is added with the customization features. It will add an action button with the reply button title text provided. | Implemented matching API that registers notification categories the user config dir (json).<br><br>On service start and shutdown the categories json is updated.<br><br>Inputs are not supported. |
| Removing Notifications | Leveraging the `UNUserNotificationCenter` API provides a ton of features including removing pending and delivered notifications.<br><br>`RemoveAllPendingNotifications`, `RemovePendingNotification`, `RemoveAllDeliveredNotifications`, and `RemoveDeliveredNotification` are available on macOS. | `RemoveAllPendingNotifications`, `RemovePendingNotification`, `RemoveAllDeliveredNotifications`, and `RemoveDeliveredNotification` are no-ops. | `RemoveNotification` removes a single notification by id, matching the identifier.<br><br>`RemoveAllPendingNotifications`, `RemovePendingNotification`, `RemoveAllDeliveredNotifications`, and `RemoveDeliveredNotification` are no-ops. |
| Notification Metadata | macOS allows custom data to be sent with the notification as `userInfo`. | Implemented a matching API by encoding and decoding the data, appending it to the one field we get back from the notification `ActivationArguments`.<br><br>For this to work we have to add the encoded data to each action for it to be returned when any action is taken. | Implemented matching API that adds a dbus variant with the data. |
| Notification Callback | `OnNotificationResponse` takes a callback thats fires when a notification is interacted with.<br>Events must be used to forward the response to the frontend. | `OnNotificationResponse` takes a callback thats fires when a notification is interacted with.<br>Events must be used to forward the response to the frontend. | `OnNotificationResponse` takes a callback thats fires when a notification is interacted with.<br>Events must be used to forward the response to the frontend. |

Notifications via `UNNotification` API is robust. The Windows Toast API supports a lot of customization options for visual appearance but interaction is not nearly as feature rich. Open to other options on Windows to close the gap!

I want to call it out here that I found this example [tiny-usernotifications-example](https://github.com/payload/tiny-usernotifications-example) to prove this could all work!

## Type of change
  
Please select the option that is relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
  
I have been staring at this a lot and tested the API through Go and JS but I am sure I have missed something. Example is included in the v3/examples dir. Help!

- [x] Windows
- [x] macOS
- [x] Linux
  - [x] Debian 12 with GNOME, Cinnamon, KDE, and more
      
If you checked Linux, please specify the distro and version.
  
## Test Configuration

```
# System

┌──────────────────────────────────────────────────┐
| Name          | MacOS                            |
| Version       | 15.3                             |
| ID            | 24D2059                          |
| Branding      | Sequoia                          |
| Platform      | darwin                           |
| Architecture  | arm64                            |
| Apple Silicon | true                             |
| CPU           | Apple M4 Max                     |
| CPU 1         | Apple M4 Max                     |
| CPU 2         | Apple M4 Max                     |
| GPU           | 32 cores, Metal Support: Metal 3 |
| Memory        | 36 GB                            |
└──────────────────────────────────────────────────┘

# Build Environment

┌─────────────────────────────────────────────────────────┐
| Wails CLI    | v3.0.0-dev                               |
| Go Version   | go1.24.1                                 |
| Revision     | 2cb0dac6484a571594931ba6eb3d198095e0436d |
| Modified     | true                                     |
| -buildmode   | exe                                      |
| -compiler    | gc                                       |
| CGO_CFLAGS   |                                          |
| CGO_CPPFLAGS |                                          |
| CGO_CXXFLAGS |                                          |
| CGO_ENABLED  | 1                                        |
| CGO_LDFLAGS  |                                          |
| GOARCH       | arm64                                    |
| GOARM64      | v8.0                                     |
| GOOS         | darwin                                   |
| vcs          | git                                      |
| vcs.modified | true                                     |
| vcs.revision | 2cb0dac6484a571594931ba6eb3d198095e0436d |
| vcs.time     | 2025-03-15T02:13:58Z                     |
└─────────────────────────────────────────────────────────┘

# Dependencies

┌────────────────────────────────────────────────────────────────────────┐
| Xcode cli tools | 2409                                                 |
| npm             | 10.9.2                                               |
| *NSIS           | Not Installed. Install with `brew install makensis`. |
|                                                                        |
└─────────────────────── * - Optional Dependency ────────────────────────┘

# Checking for issues

 SUCCESS  No issues found

# Diagnosis

 SUCCESS  Your system is ready for Wails development!
```

# Checklist:

- [x] I have updated `website/src/pages/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive, cross-platform desktop notifications system that supports both basic and interactive notifications with action buttons.
  - Enhanced support across Windows, macOS, and Linux to improve the overall user experience.

- **Documentation**
  - Added detailed guides and updated the changelog to help users set up, run, and utilize the new notification features.
  - Introduced a new "Notifications" documentation section covering initialization, sending notifications, and handling user interactions.

- **Chores**
  - Included licensing information for third-party fonts to ensure proper usage compliance.
  - Added configuration files and task management scripts for building and packaging the application across different platforms.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

popaprozac commented on v3/pkg/application/context_application_event.go:
```diff
@@ -72,6 +72,10 @@ func (c ApplicationEventContext) Filename() string {
 
+func (c *ApplicationEventContext) GetData() map[string]any {
+	return c.data
+}
```
Inadvertently left this in when my first attempt was to add a new application event. Reverted

popaprozac commented:
Hey @lyimmi 👋 

Do you think you could help with Linux support based on your prior attempt in https://github.com/wailsapp/wails/issues/1788?

popaprozac commented:
I was getting frustrated that no arbitrary data could be set on Windows Toasts. 

It's hacky but I was thinking the only thing we get back from interacting with the toast is the activation arguments on the toast itself or action button pressed. So what I am thinking is, if data is passed in with the toast, stringify and append it to the activation argument string with a separator for every defined action. When we receive the notification response we separate it back out before emitting it. See: https://github.com/wailsapp/wails/pull/4098/commits/cc760a173a29371088625060dc675db05567c029

This allows us to mirror macOS behavior and gets us metadata from the toast. Otherwise getting back the text input and action button name is kinda useless when we don't know anything else about the toast. 

I am not confident I want to keep this yet but would like some feedback.

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,305 @@
+	return true
+}
+
+// SendNotification sends a basic notification with a name, title, and body. All other options are ignored on Windows.
+// (subtitle, category id, and data are only available on macOS)
+func (ns *Service) SendNotification(options NotificationOptions) error {
+	n := toast.Notification{
+		AppID:               options.ID,
+		Title:               options.Title,
+		Body:                options.Body,
+		ActivationArguments: defaultAction,
+	}
+
+	if options.Data != nil {
+		jsonData, err := json.Marshal(options.Data)
+		if err == nil {
+			n.ActivationArguments = defaultAction + dataSeparator + strings.ReplaceAll(string(jsonData), "\"", "'")
+		}
+	}
+
+	err := n.Push()
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
I will make this more robust but will leave as is for now for us to test

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,305 @@
+		Body:                options.Body,
+		ActivationArguments: defaultAction,
+	}
+
+	for _, action := range nCategory.Actions {
+		n.Actions = append(n.Actions, toast.Action{
+			Content:   action.Title,
+			Arguments: action.ID,
+		})
+	}
+
+	if nCategory.HasReplyField {
+		n.Inputs = append(n.Inputs, toast.Input{
+			ID:          "userText",
+			Title:       nCategory.ReplyButtonTitle,
+			Placeholder: nCategory.ReplyPlaceholder,
+		})
+
+		n.Actions = append(n.Actions, toast.Action{
+			Content:   nCategory.ReplyButtonTitle,
+			Arguments: "TEXT_REPLY",
+		})
+	}
+
+	if options.Data != nil {
+		jsonData, err := json.Marshal(options.Data)
+		if err == nil {
+			n.ActivationArguments = defaultAction + dataSeparator + strings.ReplaceAll(string(jsonData), "\"", "'")
+			for index := range n.Actions {
+				n.Actions[index].Arguments = n.Actions[index].Arguments + dataSeparator + strings.ReplaceAll(string(jsonData), "\"", "'")
+			}
+		}
+	}
+
+	err := n.Push()
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
This is a workaround to get metadata from the toast when it is interacted with. Since the toast activated callback only returns the `Arguments` string of the clicked action this is a naive way of getting user provided metadata back from the toast. This means each action `Arguments` would need this appended to ensure we get it back no matter the action clicked.

popaprozac commented:
While working on getting icons working with Windows notifications with go-toast I saw the same author published [go-nativenotify](https://pkg.go.dev/git.sr.ht/~jackmordaunt/go-nativenotify) which uses their go-toast pkg on Windows, UNUserNotificationCenter on macOS, and supports Linux. I am going to see if this can be a nice drop in replacement for what I worked on since we want Linux support as well.

Otherwise I do like the way I am handling some areas and implementation details which I can write out long form after I continue playing with go-nativenotify to see if it meets our needs.

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,317 @@
+// ServiceStartup is called when the service is loaded
+// Sets an activation callback to emit an event when notifications are interacted with.
+func (ns *Service) ServiceStartup(ctx context.Context, options application.ServiceOptions) error {
+	// Need to get App Name and generate a UUID on first launch
+	//
+	// toast.SetAppData(toast.AppData{
+	// 	AppID:         "Notifications",
+	// 	GUID:          "{8F2E1A3D-C497-42B6-9E5D-72F8A169B051}",
+	// 	IconPath:      "C:\\Users\\Zach\\Development\\notifications_demo\\build\\appicon.ico",
+	// 	ActivationExe: "C:\\Users\\Zach\\Development\\notifications_demo\\bin\\Notifications.exe",
+	// })
+	//
```
Is there a nice way to get the supplied app name, preferably from the build config?

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,317 @@
+	return parts[0], parts[1]
+}
+
+// Is there a better way for me to grab this from the Wails config?
```
Similarly, good way to get the app name from the build config?

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,317 @@
+	return parts[0], parts[1]
+}
+
+// Is there a better way for me to grab this from the Wails config?
+func getExeName() string {
+	executable, err := os.Executable()
+	if err != nil {
+		return ""
+	}
+
+	return strings.TrimSuffix(filepath.Base(executable), filepath.Ext(executable))
+}
```
Similarly, good way to get the app name from the build config?

popaprozac commented:
Here is where we are at:
- macOS just needs better obj-c error handling (in progress)
- Windows is nearly there, have 2 questions: [related code here](https://github.com/popaprozac/wails/blob/050e4ef1a81cf545b1f6753ec0f022ac9281c8ab/v3/pkg/services/notifications/notifications_windows.go#L40-L47)
  1. Is there a nice way to grab the user defined app name from ideally the build config?
  2. Want to figure out out to grab the app icon properly
- Linux is stubbed out and I am playing with a promising lib that go-nativenotify is using (shout)

popaprozac commented:
Okay another update
macOS:
- switched to using channels to block on requesting and checking authorization calls
  - I forgot when I started to wire this up I was using a timeout to just kinda hack it together
  - I wanted to preserve the frontends ability to call these async methods and not rely on callbacks
- speaking of callbacks, we are going to get a warning during builds for `OnNotificationResponse` which expects a callback to deliver notification responses. How do we handle callbacks and bindings? I'm okay with it being Go only and up to the user forwarding it to frontend via event
- tried my best to account for thread safety. Can you tell I'm newish to Go??
Windows:
- I think I am happy with what I have!
Linux:
- Honestly looks promising so far on Fedora, thanks Claude
- Need to carefully walk through and test, reading freedesktop notification spec

Almost there!

leaanthony commented on v3/pkg/services/notifications/notifications.go:
```diff
@@ -0,0 +1,80 @@
+
+// OnNotificationResponse registers a callback function that will be called when
+// a notification response is received from the user.
+func (ns *Service) OnNotificationResponse(callback func(response NotificationResponse)) {
```
This should exclude it from bindings generation 👍 
```suggestion
//wails:ignore
func (ns *Service) OnNotificationResponse(callback func(response NotificationResponse)) {
```

popaprozac commented:
@leaanthony okay after a lot of reworking _I think_ we're in a good place to at least get an initial review, I've been staring at this stuff for too long. I need other eyes! 

Works across macOS, Windows, and Linux!

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,396 @@
+		cleanupChannel(id)
+		return fmt.Errorf("category registration timed out: %w", ctx.Err())
+	}
+}
```
Fixed!

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,423 @@
+	"golang.org/x/sys/windows/registry"
+)
+
+var (
+	NotificationCategories     = make(map[string]NotificationCategory)
+	notificationCategoriesLock sync.RWMutex
+	appName                    string
+	appGUID                    string
+	iconPath                   string
+)
```
Switched to initializing the `Service` in a `sync.Once` which should help here.

popaprozac replied:
The idea was to keep the API the same across platforms so they share the same minimal `Service` struct. Since those vars are not all shared I figured this was best.

Looking at how this is managed elsewhere in the codebase I see `impl` used to house different struct shapes. Can you elaborate is this can apply here? I'm tired so please help!

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,388 @@
+		cleanupChannel(id)
+		return false, fmt.Errorf("notification authorization timed out after 15s: %w", ctx.Err())
+	}
+}
```
Fixed!

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,425 @@
+			}
+			n.Actions[index].Arguments = encodedPayload
+		}
+	}
```
Fixed

popaprozac commented:
I didn't fully debug this potential issue but when I abstracted out each platform impl I found that Linux wouldn't call `ServiceStartup` if I didn't pass `ServiceOptions` along. Just a note to maybe investigate outside of here.

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,415 @@
+// Your app must be packaged and signed for this feature to work.
+func New() *Service {
+	notificationServiceOnce.Do(func() {
+		if !CheckBundleIdentifier() {
+			panic("\nError: Cannot use the notification API in development mode on macOS.\n" +
+				"Notifications require the app to be properly bundled with a bundle identifier and signed.\n" +
+				"To use the notification API on macOS:\n" +
+				"  1. Build and package your app using 'wails3 package'\n" +
+				"  2. Sign the packaged .app\n" +
+				"  3. Run the signed .app bundle")
+		}
```
I may want to change this behavior but the Obj-c code will crash and not let the app run so I thought a panic here early was more helpful. The app is not recoverable unless we stop initializing the service altogether

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,415 @@
+}
+
+// RequestNotificationAuthorization requests permission for notifications.
+// Default timeout is 3 minutes
+func (dn *darwinNotifier) RequestNotificationAuthorization() (bool, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 180*time.Second)
+	defer cancel()
+
+	id, resultCh := dn.registerChannel()
+
+	C.requestNotificationAuthorization(C.int(id))
+
+	select {
+	case result := <-resultCh:
+		return result.Success, result.Error
+	case <-ctx.Done():
+		dn.cleanupChannel(id)
+		return false, fmt.Errorf("notification authorization timed out after 15 minutes: %w", ctx.Err())
```
Fixed in latest commit

popaprozac commented:
Linux has become so unwieldy I have brought it down to the bare implementation and doing some testing. Claude initial write was good but really ran away from me. After looking at the freedesktop spec I think I can simplify

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,415 @@
+		dn.cleanupChannel(id)
+		return fmt.Errorf("category registration timed out: %w", ctx.Err())
+	}
+}
```
Fixed!

popaprozac commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,415 @@
+		dn.cleanupChannel(id)
+		return fmt.Errorf("sending notification timed out: %w", ctx.Err())
+	}
+}
```
Fixed!

popaprozac commented on v3/pkg/services/notifications/notifications_windows.go:
```diff
@@ -0,0 +1,429 @@
+	}
+
+	return n.Push()
+}
```
I can look to clean up but I am happy with this for now

popaprozac commented on v3/examples/notifications/frontend/main.js:
```diff
@@ -0,0 +1,50 @@
+import { Events } from "@wailsio/runtime";
+
+const timeElement = document.getElementById('time');
+const notificationsElement = document.getElementById('notifications');
```
The example is absolutely barebones and simple enough to ignore checks here

popaprozac commented on v3/pkg/services/notifications/notifications.go:
```diff
@@ -0,0 +1,215 @@
+	RemoveNotification(identifier string) error
+}
+
+// Service represents the notifications service
+type Service struct {
+	impl platformNotifier
+
+	// notificationResponseCallback is called when a notification result is received.
+	// Only one callback can be assigned at a time.
+	notificationResultCallback func(result NotificationResult)
+
+	callbackLock sync.RWMutex
+}
```
This is handled in each platform's implementation with the `New` method

popaprozac commented:
Updated the example to be more plug and play, commented, etc so people can test. I modified the darwin dev task to build the `.app`, sign, and open the embedded executable. From my quick testing same dev experience we want but allows us to properly use the api on macOS.

_Note: I had to also make sure the `Info.dev.plist` `CFBundleExecutable` matched the executable's name_

Recording some video so it's easy to see on each platform.

popaprozac commented:
To get dev builds working on macOS the example app modifies [this cmd](https://github.com/popaprozac/wails/blob/af5abda8b1693fb391d22a0a20a6fddfa2edd296/v3/examples/notifications/build/darwin/Taskfile.yml#L73-L80)

It seems like we could do this for all dev builds on macOS, I don't see a downside. For `build` I left it untouched, we could do the same? And for `package` you should be signing anyways to use this API and distribute the app. Thoughts?

popaprozac commented on v3/pkg/services/notifications/notifications_windows_icon.go:
Maybe this moves to `pkg/w32/icon.go`?

leaanthony replied:
Yeah, I think so. There's already very similar code there.

leaanthony commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,412 @@
+package notifications
+
+/*
+#cgo CFLAGS: -mmacosx-version-min=11.0 -x objective-c
```
I think this will be problematic because it will fail to compile on systems < 11. 
I believe the right approach is to do something like this: 
```
/*
#cgo CFLAGS: -x objective-c
#cgo LDFLAGS: -framework Foundation -framework UserNotifications

#import <Foundation/Foundation.h>

#if __MAC_OS_X_VERSION_MAX_ALLOWED >= 110000
#import <UserNotifications/UserNotifications.h>
#endif

BOOL canSendNotifications() {
    // Using @available for runtime check
    if (@available(macOS 11.0, *)) {
        return YES;
    } else {
        return NO;
    }
}

BOOL sendNotificationIfAvailable(const char* title, const char* body) {
    if (@available(macOS 11.0, *)) {
        // Implementation using UserNotifications framework
        UNUserNotificationCenter *center = [UNUserNotificationCenter currentNotificationCenter];
        
        UNMutableNotificationContent *content = [[UNMutableNotificationContent alloc] init];
        content.title = [NSString stringWithUTF8String:title];
        content.body = [NSString stringWithUTF8String:body];
        
        UNNotificationRequest *request = [UNNotificationRequest requestWithIdentifier:[[NSUUID UUID] UUIDString]
                                                                              content:content
                                                                              trigger:nil];
        
        [center addNotificationRequest:request withCompletionHandler:nil];
        return YES;
    } else {
        // Not available on this OS version
        NSLog(@"Notification API requires macOS 11.0 or later");
        return NO;
    }
}
*/```
We can then do things like:
```go
// IsNotificationAvailable checks if notifications are available on this system
func IsNotificationAvailable() bool {
    return bool(C.canSendNotifications())
}``` and give the developer the option of what to do if they are not available.

leaanthony commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,412 @@
+func New() *Service {
+	notificationServiceOnce.Do(func() {
+		if !CheckBundleIdentifier() {
+			panic("\nError: Cannot use the notification API in development mode on macOS.\n" +
```
I think it's better if we return this as an error rather than panic. A corrupt bundle at runtime may cause it so the developer may want to handle it as a crash and save some information or show a dialog or something.

popaprozac replied:
Makes sense and, related to checking compatibility and return an err,  do we want to return on calling `New` or in `ServiceStartup`? I don't really see a precedent for return `(*Service, error)`

leaanthony commented on v3/pkg/services/notifications/notifications_darwin.go:
```diff
@@ -0,0 +1,412 @@
+		return
+	}
+
+	dn, ok := ns.impl.(*darwinNotifier)
```
What do you think about having the impl as an interface? That way you wouldn't need to type cast it like this. Just a thought.

popaprozac replied:
Yeah makes sense. My brain is still conforming to Go norms.
```
type ChannelHandler interface {
	GetChannel(id int) (chan notificationChannel, bool)
}

var _ ChannelHandler = (*darwinNotifier)(nil)
```
Something like that?

leaanthony commented on v3/examples/notifications/Taskfile.yml:
```diff
@@ -0,0 +1,34 @@
+    cmds:
+      - task: "{{OS}}:package"
+
+  run:
```
Is it worth bringing in the "run" changes from the updated dev experience PR?

lujihong commented:
windows下系统通知依赖[go-toast](https://pkg.go.dev/git.sr.ht/~jackmordaunt/go-toast/v2@v2.0.3)，但是我使用遇到问题，我在项目中使用了 uiautomation，然而[go-toast](https://pkg.go.dev/git.sr.ht/~jackmordaunt/go-toast/v2@v2.0.3)中的initialize方法中使用了ole.RoInitialize(1)，会和 ole.CoInitializeEx 冲突导致uiautomation无法正常使用。

我尝试修复，对 git.sr.ht\~jackmordaunt\go-toast\v2@v2.0.3\wintoast\bind_windows.go 文件修改initialize方法及调用逻辑：

```
func initialize() (func(), error) {
	runtime.LockOSThread()

	err := ole.RoInitialize(1)
	if err != nil {
		runtime.UnlockOSThread()
		if oleErr, ok := err.(*ole.OleError); ok && oleErr.Code() == 0x80010106 {
			return nil, fmt.Errorf("RoInitialize: RPC_E_CHANGED_MODE — thread previously initialized with different mode")
		}
		return nil, fmt.Errorf("RoInitialize: %w", err)
	}

	release := func() {
		ole.RoUninitialize()
		runtime.UnlockOSThread()
	}
	return release, nil
}

func pushCOM(appID, xml string) (err error) {
	//.........

	release, err := initialize()
	if err != nil {
		return err
	}
	defer release()

	//.........
}
```
